### PR TITLE
Migrate to pytest

### PR DESCRIPTION
**Pull request by bashtage**
Switch testing to pytest 

Lots of testing clean up
Some fixes to ensure tests are picked up by pytest

### CODE REVIEW

**Comment by bashtage** [comment]
The reduction in coverage is due to coverage on pytest counting files that have 0 lines executed while coverage with nose ignores these files.  These are mostly in sandbox.

**Comment by josef-pkt** [comment]
as a time line here: I think we should make the switch soonish (in 3 weeks, I'm still on vacation for two weeks with only short attention spans)

Pauli has started the transition to pytest for scipy, so we will get additional information and examples if needed.

The merged VECM PR added a large number of yield based unit tests. I think we will convert those to non-yield asserts once we are reasonably sure we don't need to do much debugging anymore (#3775)

**Comment by bashtage** [comment]
I think the right thing to do it so get this in since it is correct right now and allows wither pytest or nose.

I have carefully checked coverage file-by-file and there is nothing not covered by pytest that is by nose (except a tiny number of lines which are simply accounting differences since nose executes classes while pytest does not).

Once this is in the yields can be changed to parameterize (this isn't necessary, one could pin pytest<4).

**Comment by josef-pkt** [comment]
crash in matplotlib are back using pytest

M:\>nosetests M:\josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodel
s\graphics
results in 
Ran 117 tests in 39.610s
OK

```M:\>pytest M:\josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\g
raphics
============================= test session starts =============================
platform win32 -- Python 3.4.4, pytest-3.0.3, py-1.4.31, pluggy-0.4.0
rootdir: M:\josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr, inifile:
collected 117 items

josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_boxplots.py .
josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_correlation.py ..
josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_dotplot.py .
josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_factorplots.py .......
josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_functional.py ...
josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_gofplots.py .......................................................
josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_mosaicplot.py ...............
josef_new\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\
test_regressionplots.py ............
```
segfault

the previous fix for those segfaults with nosetests was to change the unit tests so that plot figures are closed.
Maybe pytest doesn't close them soon enough.

**Comment by josef-pkt** [comment]
fix for graphics segfault is in #3844

File: statsmodels/__init__.py [hunk 1]
```diff
@@ -25,55 +26,29 @@
...
+    def __call__(self, extra_args=None):
+        try:
+            import pytest
+            if not LooseVersion(pytest.__version__) >= LooseVersion('3.0'):
+                raise ImportError
+            extra_args = [] if extra_args is None else extra_args
+            cmd = [self.package_path] + extra_args
+            print('Running pytest '+ ' '.join(cmd))
+            errno = pytest.main(cmd)
+            sys.exit(errno)
```
Comment (josef-pkt) [inline statsmodels/__init__.py]: I think it shouldn't exit, the same way as before. 
i.e. I was running different tests in the same python session

**Comment by josef-pkt** [comment]
I didn't find any other problems

I think the default runner `statsmodels.test()` should include --tb=short  --disable-pytest-warnings, and not sys.exit. 

For travis we can adjust the options later again. I assume we want to switch back to verbose mode, but that can be decided after we see a few failures in PRs on Travis.

Otherwise, I think we can merge this now after rebase, and then fix PRs as they get rebased on master.

Does the test runner on TravisCI stop immediately if there is a test that is pytest incompatible or show some warnings at the end?

**Comment by josef-pkt** [comment]
Does the pytest output filter and suppress RuntimeWarnings? I see almost only our own warnings with pytest in contrast to nosetests
e.g.
https://travis-ci.org/statsmodels/statsmodels/jobs/259045322

**Comment by bashtage** [comment]
There are FutureWarnings from NumPy  showing in that build.  Should there be RuntimeWarnings?

**Comment by josef-pkt** [comment]
running with nosetests shows a large number of RuntimeWarnings, e.g. for python 3.6 run
https://travis-ci.org/statsmodels/statsmodels/jobs/259045323

many Warnings for invalid values, zero division, ... in math computations,  (including a lot of invalid values noise from scipy.stats)
I don't know how much of those are noise that hasn't been silenced and how much might hide more serious issues

**Comment by bashtage** [comment]
@josef-pkt The warning reports are back. There is a bug with the multi threaded runner that suppressed them. There are many yield warnings.  Once you merge this, I'mm start working through those.

**Comment by josef-pkt** [comment]
I will merge this after travis reports back. Then we can continue with master and get some rebased PRs to see how well it works.

For the yield test it might be easiest to just convert the `yield` to regular numpy asserts,, so we don't loose nosetests compatibility right away. VECM looks good and the other yield tests are older so I don't expect that we need them for debugging right now.

**Comment by josef-pkt** [comment]
I still don't see the usual RuntimeWarning
https://travis-ci.org/statsmodels/statsmodels/jobs/259394059 has one RuntimeWarning that is issued in statsmodels.
So it looks to me that pytest is filtering RuntimeWarnings that don't come from statsmodels, e.g. the numpy and scipy float operation and linalg warnings.

Nevertheless, I'm merging this now.
@bashtage  Thanks for this big refactor

**Pull request merged**